### ISSUE TITLE
Dynamic update of grid options after the grid is loaded

### ISSUE THREAD

**Issue by satishtatavarty** [ag-grid/ag-grid#2149]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[X] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

is it possible to do dynamic update of grid options after the grid is loaded
**Expected behavior**

dynamic update of grid options after the grid is loaded
**Minimal reproduction of the problem with instructions**

**What is the motivation / use case for changing the behavior?**

We have a singleton instance across the app and there's need for the grid to adapt based on sundry use cases with different behaviors.
**Please tell us about your environment:**

* **ag-Grid version:** X.X.X

15.0
* **Browser:**

Chrome
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
ES6

**Comment by JbRezidu** [comment]
Hi, 
I'm also looking for something like this.
I want to change the stopEditingWhenGridLosesFocus option depending on the cell editor that is used but I'm not sure that this is possible...

**Comment by enchanter5525** [comment]
@jbrezidu did you solve the issue that you faced?
I m facing the similiar one. I want to change stopeditingwhengridlosesfocus dynamically

**Comment by lynx-r** [comment]
I dit it like this:

```
...
  const gridOptions = useMemo(() => GRID_OPTIONS[type], [type])

  const AgGridComponent = useCallback(
    () => AgGridProxy({gridOptions, onGridReady, rowData}),
    [gridOptions, onGridReady, rowData])

  return (
    <div className="ag-theme-alpine w-100 h-100">
      <AgGridComponent/>
    </div>
  )
...
```

**Comment by alessandrohbnr** [comment]
@lynx-r What is this AgGridProxy? The AgGrid component imported?

**Comment by lynx-r** [comment]
@alessandrohbnr yes. It contains AgGrid table. Actually `useCallback` creates a new AgGrid table instance on every change of gridOptions or rowData.

**Comment by alessandrohbnr** [comment]
Yeah but I tried to import AgGrid like this:
`import { AgGridProxy } from 'ag-grid-react'`
and it was not found, am I missing something?

**Comment by lynx-r** [comment]
No, it is the custom wrap around AgGrid table component.

**Comment by alessandrohbnr** [comment]
Do you have any example of this wrap?
Just to let you know I'm trying to make some columns editable for all table when the user clicks on an edit button. I'm thinking to use the `suppressClickEdit` to do that change with some state flag and updating the gridOptions dinamically.

**Comment by lynx-r** [comment]
Try this idea:
```
// custom-aggrid.jsx
export default (columnDefs, rowData, onGridReady) => {
  return <AgGridReact
                // properties
                columnDefs={columnDefs}
                rowData={rowData}

                // events
                onGridReady={onGridReady}>
   </AgGridReact>
}

// aggrid-component.jsx
// COLUMN_DEFS a dict with non editable and editable grid options. Like COLUMN_DEFS = {editable: [...], nonEditable: [...]}
// type ColumnDefsType = 'editable' | 'nonEditable'
  let type: ColumnDefsType = 'nonEditable'
  const columnDefs = useMemo(() => COLUMN_DEFS[type], [type])

  const AgGridComponent = useCallback(
    () => CustomAggrid(columnDefs, rowData, onGridReady),
    [gridOptions, onGridReady, rowData])

  return (
    <button onClick="type = 'editable'">Edit table</button>
    <div className="ag-theme-alpine w-100 h-100">
      <AgGridComponent/>
    </div>
  )
```
So when type is changed the columnDefs `useMemo` triggers and then AgGridComponent `useCallback` triggers and recreates a new table with new columnDefs.

**Comment by lynx-r** [comment]
Also you can use a function instead of COLUMN_DEFS like
const getColumnDefs = (isEditable: boolean) => [...{..., editable: isEditable}]

**Comment by StephenCooper** [comment]
To provide an update on this ticket as it is getting some attention. We have an ongoing track to make more and more of the grid properties reactive so that they can be updated without re-creating the entire grid. This can be tracked under **AG-5154 Make properties/callbacks reactive**

`ColumnDefs` is already reactive so you can provide updated column definitions to an existing grid and it will update as required without re-creating the entire grid. Please see https://ag-grid.com/javascript-data-grid/column-updating-definitions/

As for other properties we are working on them on a case by case basis to make them reactive.

**Comment by vishalpawar048** [comment]
Hi @StephenCooper ,

I tried to find AG-5154 but couldn't find it [here](https://www.ag-grid.com/changelog/).

I am also trying to change domLayout in gridoption as per the data to adjust the height of the grid.
